src/runner.py

```diff
@@ -0,0 +1,576 @@
+
+    @retry(tries=5, delay=1, local_logger=logger)
+    def _get_resource_profile(self, resources: VirtualMachineResources) -> str:
+        """Get the LXD profile name of given resource limit.
+
+        Args:
+            resources: Resources limit of the runner instance.
+
+        Raises:
+            RunnerError: Unable to create the profile on LXD.
+
+        Returns:
+            str: Name of the profile for the given resource limit.
+        """
+        # Ensure the resource profile exists.
+        profile_name = f"cpu-{resources.cpu}-mem-{resources.memory}-disk-{resources.disk}"
+        if not self._clients.lxd.profiles.exists(profile_name):
+            logger.info("Creating LXD profile for resource usage.")
+            try:
+                resource_profile_config = {
+                    "limits.cpu": str(resources.cpu),
+                    "limits.memory": resources.memory,
+                }
+                resource_profile_devices = {
+                    "root": {
+                        "path": "/",
+                        "pool": "default",
+                        "type": "disk",
+                        "size": resources.disk,
+                    }
+                }
+                self._clients.lxd.profiles.create(
+                    profile_name, resource_profile_config, resource_profile_devices
+                )
+            except LxdError as error:
+                logger.error(error)
+                raise RunnerError(
+                    "Resources were not provided in the correct format, check the juju config for "
+                    "cpu, memory and disk."
+                ) from error
+
+            # Verify the action is successful.
+            if not self._clients.lxd.profiles.exists(profile_name):
+                raise RunnerError(f"Unable to create {profile_name} LXD profile")
+        else:
+            logger.info("Found existing LXD profile for resource usage.")
+
+        return profile_name
+
+    @retry(tries=5, delay=1, local_logger=logger)
+    def _start_instance(self) -> None:
+        """Start an instance and wait for it to boot.
+
+        Args:
+            reconcile_interval: Time in seconds of period between each reconciliation.
+        """
+        if self.instance is None:
+            raise RunnerError("Runner operation called prior to runner creation.")
+
+        logger.info("Starting LXD instance for runner: %s", self.config.name)
+
+        # Setting `wait=True` only ensure the instance has begin to boot up.
+        self.instance.start(wait=True)
+
+    @retry(tries=5, delay=30, local_logger=logger)
+    def _wait_boot_up(self) -> None:
+        if self.instance is None:
+            raise RunnerError("Runner operation called prior to runner creation.")
+
+        # Wait for the instance to finish to boot up and network to be up.
+        self.instance.execute(["/usr/bin/who"])
+        self.instance.execute(["/usr/bin/nslookup", "github.com"])
```

Comment:
The exit code of the execution is not checked, same for all occurrences of `self.instance.execute` in the charm.

Reply:
Major comments